addons/resource_distribution_controller.go

```diff
@@ -0,0 +1,161 @@
+			return object.GetName() == utils.StorageClientMappingConfigMapName && object.GetNamespace() == r.CurrentNamespace
+		}))).
+		Watches(&templatev1.Template{}, handler.EnqueueRequestsFromMapFunc(func(_ context.Context, _ client.Object) []ctrl.Request {
+			return []ctrl.Request{reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: r.CurrentNamespace,
+					Name:      utils.StorageClientMappingConfigMapName,
+				},
+			}}
+		}), builder.WithPredicates(labelSelectorPredicate)).
+		WithEventFilter(predicate.NewPredicateFuncs(func(object client.Object) bool {
+			return object.GetNamespace() == r.CurrentNamespace
+		})).
+		Complete(r)
+}
+
+func (r *ResourceDistributionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger.With("Request", req)
+	logger.Info("Distributing resources to all StorageConsumers.")
+
+	var err error
+
+	var addonDeletionlock corev1.ConfigMap
+	err = r.SpokeClient.Get(ctx, types.NamespacedName{Namespace: r.CurrentNamespace, Name: AddonDeletionlockName}, &addonDeletionlock)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !slices.Contains(addonDeletionlock.GetFinalizers(), ResourceDistributionFinalizer) {
+		logger.Info("Adding finalizer to addonDeletionlock", "finalizer", ResourceDistributionFinalizer)
+		addonDeletionlock.Finalizers = append(addonDeletionlock.Finalizers, ResourceDistributionFinalizer)
+		if err := r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	addVRCToConsumers := []*ocsv1alpha1.StorageConsumer{}
+	storageConsumerList := ocsv1alpha1.StorageConsumerList{}
+	templateList := templatev1.TemplateList{}
+
+	err = r.SpokeClient.List(ctx, &storageConsumerList, client.InNamespace(r.CurrentNamespace))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(storageConsumerList.Items) == 0 {
+		logger.Info("No StorageConsumer to add/remove resources.")
+		if slices.Contains(addonDeletionlock.GetFinalizers(), ResourceDistributionFinalizer) {
+			logger.Info("Removing finalizer from addonDeletionlock", "finalizer", ResourceDistributionFinalizer)
+			addonDeletionlock.Finalizers = utils.RemoveString(addonDeletionlock.Finalizers, ResourceDistributionFinalizer)
+			if err := r.SpokeClient.Update(ctx, &addonDeletionlock); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	err = r.SpokeClient.List(ctx, &templateList, client.InNamespace(r.CurrentNamespace), client.MatchingLabels{
+		utils.CreatedByLabelKey: "odf-multicluster-orchestrator",
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to distribute resources to StorageConsumers: %w", err)
+	}
+	if len(templateList.Items) == 0 {
+		for _, consumer := range storageConsumerList.Items {
+			consumer.Spec.VolumeReplicationClasses = []ocsv1alpha1.VolumeReplicationClassSpec{}
```

Comment:
The user doesn't create any VRC for distribution, but if they do this will delete the user created VRC's as well

Reply:
That is expected. I tried allowing list merges, but then it becomes impossible to track who is responsible for which VRCs. So, I chose to take control of entire list.
Users can either let MCO manage it, or manage it themselves.
If MCO manages it, users can still create the template and MCO will add it to this list. So there is a way out for them.
If user manages it, MCO can not touch it.

Reply:
I think this is problematic, mco shouldn't be deleting what mco didn't create

Reply:
@umangachapagain From a high level point of view, if MCO added something to the list then it needs to remove it but not touch anything anyone else added to the list. If there is a technical difficulty to achieve that let's work together to come up with a solution.

Reply:
@nb-ohad @rewantsoni I am open to suggestions on how this can be correctly implemented.